### PR TITLE
Yul codegen for member access to type type

### PR DESCRIPTION
**Pull request by chriseth**
Depends on #8684 
Part of #8343.

### CODE REVIEW

**Comment by chriseth** [comment]
I think some more tests could be activated. Please take over this part :)

**Comment by chriseth** [comment]
Ready for review.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -985,12 +985,72 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
+					break;
+				case FunctionType::Kind::Internal:
+					if (auto const* function = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration))
+						define(_memberAccess) << to_string(function->id()) << "\n";
+					else
+						solAssert(false, "Function not found in member access");
+					break;
+				case FunctionType::Kind::Event:
+					if (!dynamic_cast<EventDefinition const*>(_memberAccess.annotation().referencedDeclaration))
+						solAssert(false, "event not found");
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
					solAssert(dynamic_cast<EventDefinition const*>(_memberAccess.annotation().referencedDeclaration), "event not found");
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -985,12 +985,72 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
-		solUnimplementedAssert(funType, "");
-		solUnimplementedAssert(funType->kind() == FunctionType::Kind::Declaration, "");
-		// Nothing to do for declaration.
+		Type const& actualType = *dynamic_cast<TypeType const&>(
+			*_memberAccess.expression().annotation().type
+		).actualType();
+
+		if (actualType.category() == Type::Category::Contract)
+		{
+			if (auto variable = dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I think we should stick to extracting `const`, `&` and `*` from auto, or at least be consistent with it.
Down few lines there is usage like `auto const*`

**Comment by mijovic** [comment]
Looks good to me, except few small style suggestions

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -985,12 +985,74 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
-		// Nothing to do for declaration.
+		Type const& actualType = *dynamic_cast<TypeType const&>(
+			*_memberAccess.expression().annotation().type
+		).actualType();
+
+		if (actualType.category() == Type::Category::Contract)
+		{
+			if (auto const* variable = dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+				handleVariableReference(*variable, _memberAccess);
+			else if (auto funType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			else if (auto const* funType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
```

**Review by leonardoalt — DISMISSED** [review]
Just this one tiny comment, otherwise looks good!

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**